src/charm.py

```diff
@@ -74,16 +75,29 @@ def sha256(hashable) -> str:
     return hashlib.sha256(hashable).hexdigest()
 
 
+def pull(container, path, *, default: Optional[str] = None) -> Optional[str]:
+    """Pull file from container (without raising FileNotFoundError).
```

Comment:
Other than `default` never being used (and its usage probably breaking the logic of the other places this is called anyway), this has essentially the same question as "why was `push` moved to the top level"? 

In the worst case, this could be a class-level method named `try_pull` or `get_file_contents` with something like:

```python
def get_file_contents(self, path) -> Optional[str]:
  try:
    return self.container.pull(path).read()
  except pebble.PathError:
    logger.debug(...)
    return None
```

Or use `contextlib.suppress`. But Pebble should be returning a `PathError` if it doesn't exist (there's [a test for that, too](https://github.com/canonical/operator/blob/7faa5a2451e4531492d4513feaea0617512f6e9d/test/test_pebble.py#L1893), but the `client` is a Mock, and if Pebble doesn't really do that at runtime, we should raise it with Ben). A `FileNotFoundError` means it's getting [here](https://github.com/canonical/operator/blob/main/ops/pebble.py#L2449), and Pebble [should be raising before it gets there](https://github.com/canonical/operator/blob/7faa5a2451e4531492d4513feaea0617512f6e9d/ops/pebble.py#L1824).

All of that aside, extracting these methods makes the flow of the code much harder for no clear benefit

Reply:
Yes, I'm not too excited about those free functions myself.
Can easily change but here's my thinking.

1. Good point about default. Removed.
2. Thanks for the PathError tip, https://github.com/canonical/operator/issues/896.
3. Currently we only pull hash files. It will always be `utf-8` and I think we always want None back. Repeating a contextlib is fine but like this it's less distractions while reading the main code.
4. I don't think the "jumping around" argument is valid - we jump around all the time everywhere. Also in our charming context it's pretty obvious what "pull" and "push" do.
5. I do not see justification to make it a method (nor a staticmethod) (we have lots of such "methods" already but it's not a reason to continue doing so). What's wrong with good old free functions?
6. `try_pull` or `pull_or_none` are fine. Just `pull` also seems fine.

Reply:
Converted pull, push to methods ([diff](https://github.com/canonical/prometheus-k8s-operator/pull/431/commits/2b5da290969e5e297dad9d0933e99a5538f4933c)).

Reply:
Hopefully the OF bug will make it nicer. It looks like it's missing some change in teh return from the Pebble API.

Part of the thing here is around scoping and least responsibility. It's not always easy to get used to a strange/new codebase, and despite the fact that we do "jump around" (during larger patches/refactors, I do at least try to shuffle similar methods near each other, so all of the "alert" stuff is near the "alert" stuff), but somewhat the problem is that it is _not_ obvious what "push" and "pull" do. 

For any other author used to reading/writing charm code, or looking at the docs, `push` and `pull` here do not behave like `ops.model.Container.[push|pull]`. The behavior is different, even if only slightly, rather than having a method which is a tightly scoped as possible (for the case of "what if a hash value is empty?") and named to reflect that. 

_Currently_ we only pull hash files, but that may change in the future, and when/if it does, we must contend with either a "`push|pull` implementation which is specific to this file and not like the rest of the ecosystem", using this `push|pull` for hash files and the "normal" `push|pull` elsewhere, the author who writes the patch to "pull" whatever files are needed then refactoring this out (since `ops` already does it), and so on.

The argument to make it a method is essentially twofold: that the "domain" which owns a container is the instance of the charm, and that the only real "consumer" of that method is also the charm. This somewhat applies to making things `@staticmethod` also. We are now at a point where external authors are reading our code, and using it, and Hyrum's law applies -- every casual change is potentially one which we must live with until we come to a point in time where we _can_ break the API.

There's nothing wrong with top-level functions or `@staticmethod` if they're actually intended to be used by external users, or don't require peering inside objects, or are otherwise visible. Pulling out a method which is only used by one class to the top level (or making a method on a class which isn't generally reusable a `@staticmethod` because it doesn't actually mutate state and it's possible) doesn't end up being good encapsulation, and providing "free" methods which don't require an instance of the object is somewhat asking for them to be used.

Put differently, [this method](https://github.com/canonical/prometheus-k8s-operator/blob/c21baa67cbf7841860793441c602af9c59855650/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L455) is only called from a single place (in our code), from a private method. Whatever notion of idempotency or state machine we have inside our charms is often in controlling the "flow" of data and transformations from event->handler->read relation data->transform->ingest by workload. 

It's ok to have this where it is. It is idempotent, in that the "worst thing" that happens is pretty much `job_name` getting more strings tacked onto it. It's nice and clean. But it's also true that now we can't ever change it, because we don't know who may be using it, and if we have a need to do some other kind of manipulation, we'll need to put it as a private instance method we control anyway.  

There will, of course, be times to do this either way, and times where that tradeoff is acceptable, but I don't think that we should do it just because we can, because we _cannot_ take it back in those circumstances. Similarly, if we want to assume that every time we're looking for _hashes_ it should have some set of options (`make_dirs=True, encoding="utf8"`), we could make a `functools.partial` and apply whatever options, or narrow it to that scope. 

`push` and `pull` should not be restrictive enough that it's _always_ UTF-8 text, or `None` (and that the author won't want an exception to treat it like "normal" `ops.model.Container.pull`). When secret support lands, there _are_ some secret/authentication methods which are binary `.p12` by default, and this will just block them out. Or use "regular" `push()`, and now we have both paradigms in one charm with the same name, or teach this one to accept `*args, **kwargs` and forward them, or ...

This particular logic around "what to do if a file is not found, what particular options to push a hash with" could be cleanly isolated to this functionality.

I'm not trying to be too harsh or argumentative on the review here, but we need to be careful about making very large changes without a good reason

Reply:
The scope / "blast radius" of these functions is just the charm.
We have many methods that make assumptions that make them not as general as they could be.
And we modify those methods of ours per need.
These push/pull are no exception.
Taking Jose's recommendation to make them "private" with `_`.

Reply:
I'm not going to hold this patch up, but even if the scope/"blast radius" of these function is just the charm, it is still "now we have two almost identically-named ways to push/pull in the same charm codebase" (if someone writes a later patch which uses the "normal" push/pull) or "we have a push/pull which doesn't behave the same way as the 'standard' and requires special handling" rather than keeping it isolated to the one use case it's actually needed for and re-evaluating later.

Again, once something is public (top-level, `PrometheusConfig....`), we're kind of "locked in" to a public API, for better or worse, and we need to be far more careful with it.

I am not convinced that "we already have methods which make assumptions" is a good argument for adding more of them.

What is the downside to keeping these methods scoped to/named for their solitary function (hash push/pulling), whether public instance methods or private/

Reply:
No, I don't think that `_pull` is public and binding at all (I meant that more in reference to the `PrometheusConfig...` methods). I'm more worried, for this change that if we get to needing [binaries from secrets](https://github.com/canonical/operator/issues/869) that there will be a confusing blend of `_pull` (with its inherent assumptions) and `self.charm.container.pull` (where we need/want binary).

We can defer that as a future case -- not blocking this patch, but it's something worth considering. Even if we don't have a use case right now, it's entirely plausible that we actually will have a use case or need for the "normal" `push|pull` interface in the future, and we may as well leave the door open for that (in all of our charms).

Reply:
Thanks for the clarification!
Agreed about secrets.